Topics/System_Design/Intro_To_Microservices/Intro_To_Microservices.md

```diff
@@ -0,0 +1,165 @@
+6. [Tools and Technologies](#common-tech)
+7. [Best Practices](#best-practices)
+8. [Real-World Examples](#examples)
+
+
+
+## 1. Overview of Microservices <a name="overview"></a>
+
+In the realm of CSC301H1: Introduction to Software Engineering, understanding modern software architecture is crucial. Microservices architecture, a cornerstone of contemporary software development, is a concept that aligns seamlessly with the principles taught in this course.
+
+Microservices architecture presents a paradigm shift in how software systems are designed and developed. Unlike traditional monolithic architectures, where applications are built as a single, tightly integrated unit, microservices architecture emphasizes breaking down complex systems into smaller, independently deployable services.
+
+For CSC301 students, grasping the significance of microservices architecture entails understanding its foundational principles and how they apply to real-world software engineering practices. Microservices enable teams to work more efficiently by allowing them to focus on developing and maintaining small, autonomous services. This aligns with the agile development methods taught in CSC301, where adaptability, collaboration, and rapid iteration are paramount.
+
+In the context of CSC301, microservices architecture offers several advantages. It promotes scalability, allowing teams to scale individual services based on demand, a concept directly relevant to discussions on estimation and prioritization. Moreover, the autonomy of microservices facilitates teamwork skills, as teams can work independently on different services while ensuring seamless integration through well-defined APIs.
+
+However, adopting microservices architecture also introduces challenges, such as managing the complexity of a distributed system and ensuring effective communication between services. These challenges underscore the importance of basic modeling, design patterns, and refactoring, topics covered in CSC301.
+
+Overall, understanding microservices architecture within the context of CSC301 provides students with valuable insights into modern software engineering practices. By delving into the principles, advantages, and challenges of microservices, CSC301 students gain a solid foundation in designing scalable, resilient, and maintainable software systems—a cornerstone of professional software engineering.
+
+## 2. Advantages and Disadvantages of Microservices Architecture <a name="adv-dis"></a>
+
+Microservices architecture presents several advantages and challenges that CSC301 students should consider when evaluating its adoption for software development projects.
+
+**Advantages:**
+
+1. **Scalability**: Microservices enable horizontal scaling, allowing organizations to handle increased traffic by deploying multiple instances of individual services. This scalability is crucial for projects with varying demand, as discussed in CSC301's emphasis on agile development methods.
```

Comment:
Good usage of the term "horizontal scaling". I am not sure if we touched upon it during our lectures. A more layman explanation would have been better. Such as "horizontal scaling refers to the instance, where multiple servers are deployed which are handling similar tasks".

Reply:
An example would indeed be helpful now that I think about it!